File: packages/sdk-python/src/neko_oauth2_sdk/webhook.py
import hashlib
import hmac
import json
from dataclasses import dataclass
from typing import Any, Optional, Union


@dataclass
class WebhookEvent:
    event: str
    data: Any
    timestamp: str


class InvalidWebhookSignature(Exception):
    pass


def verify_webhook_signature(raw_body: Union[bytes, str], signature_header: Optional[str], secret: str) -> bool:
    """Matches the server's own signing scheme exactly: `X-Neko-Signature:
    sha256=<hex hmac-sha256 of the raw request body>`. `raw_body` must be the
    exact bytes the server sent -- re-serializing after parsing JSON can
    differ enough (key order, whitespace) to fail verification even for
    "the same" content, so callers must capture the raw body before parsing.
    """
    if not signature_header:
        return False
    scheme, _, signature = signature_header.partition("=")
    if scheme != "sha256" or not signature:
        return False

    if isinstance(raw_body, str):
        raw_body = raw_body.encode("utf-8")

    expected = hmac.new(secret.encode("utf-8"), raw_body, hashlib.sha256).hexdigest()
    try:
        expected_bytes = bytes.fromhex(expected)
        provided_bytes = bytes.fromhex(signature)
    except ValueError:
        return False
    if len(expected_bytes) != len(provided_bytes):
        return False
    return hmac.compare_digest(expected_bytes, provided_bytes)


def parse_webhook_payload(raw_body: Union[bytes, str]) -> WebhookEvent:
    data = json.loads(raw_body)
    return WebhookEvent(event=data["event"], data=data.get("data"), timestamp=data.get("timestamp", ""))


def verify_and_parse_webhook(raw_body: Union[bytes, str], signature_header: Optional[str], secret: str) -> WebhookEvent:
    if not verify_webhook_signature(raw_body, signature_header, secret):
        raise InvalidWebhookSignature("invalid webhook signature")
    return parse_webhook_payload(raw_body)

File: packages/sdk-go/webhook.go
package neko

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"strings"
)

// WebhookEvent mirrors the server's payload shape exactly:
// { "event": "...", "data": {...}, "timestamp": "..." }.
type WebhookEvent struct {
	Event     string          `json:"event"`
	Data      json.RawMessage `json:"data"`
	Timestamp string          `json:"timestamp"`
}

var ErrInvalidWebhookSignature = errors.New("invalid webhook signature")

// VerifyWebhookSignature matches the server's own signing scheme exactly:
// `X-Neko-Signature: sha256=<hex hmac-sha256 of the raw request body>`.
// rawBody must be the exact bytes the server sent -- re-serializing JSON
// after parsing can differ enough (key order, whitespace) to fail
// verification even for "the same" content.
func VerifyWebhookSignature(rawBody []byte, signatureHeader string, secret string) bool {
	if signatureHeader == "" {
		return false
	}
	parts := strings.SplitN(signatureHeader, "=", 2)
	if len(parts) != 2 || parts[0] != "sha256" {
		return false
	}

	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(rawBody)
	expected := mac.Sum(nil)

	provided, err := hex.DecodeString(parts[1])
	if err != nil {
		return false
	}
	return hmac.Equal(expected, provided)
}

func ParseWebhookPayload(rawBody []byte) (*WebhookEvent, error) {
	var event WebhookEvent
	if err := json.Unmarshal(rawBody, &event); err != nil {
		return nil, err
	}
	return &event, nil
}

func VerifyAndParseWebhook(rawBody []byte, signatureHeader string, secret string) (*WebhookEvent, error) {
	if !VerifyWebhookSignature(rawBody, signatureHeader, secret) {
		return nil, ErrInvalidWebhookSignature
	}
	return ParseWebhookPayload(rawBody)
}
